### PR TITLE
Request agnostic

### PR DESCRIPTION
**Pull request by lordrhodos**
This PR makes the library (partly) request agnostic by adding support for the following request types:

* PSR-7 requests
* Symfony requests (symfony/http-foundation)
* Guzzle5 requests (Guzzl6 implements PSR-7 interfaces)

In addition this PR supports passing in a custom `HmacDataTransformerInterface` implementation if the format of the data string needs to be customized.

This PR is based on the branch of PR https://github.com/starweb/starlit-request-authenticator/pull/3

### CODE REVIEW

File: src/Hmac/Adapter/SymfonyRequestAdapter.php [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+     * helper method to build a uri without reordering url parameters in a query string alphabetically
+     * which the interal Request::getUri function does (called normalize)
+     *
+     * @see SymfonyRequest::getUri()
+     * @see SymfonyRequest::getQueryString()
+     * @see SymfonyRequest::normalizeQueryString()
+     */
+    public function getUri(): string
+    {
+        $qs = $this->request->server->get('QUERY_STRING');
```
Comment (felixsand) [inline src/Hmac/Adapter/SymfonyRequestAdapter.php]: Maybe rename `$qs` to `$queryString` instead in order to make it's use clearer?

File: tests/Hmac/HmacAuthenticatorTest.php [hunk 2]
```diff
@@ -30,22 +31,10 @@ public function testConstructor(): void
         $this->assertInstanceOf(AuthenticatorInterface::class, $this->authenticator);
     }
 
-    /**
-     * @covers \Starlit\Request\Authenticator\Hmac\HmacAuthenticator::__construct()
-     */
-    public function testConstructorWithEmptyStringAsSecretThrowsException()
```
Comment (felixsand) [inline tests/Hmac/HmacAuthenticatorTest.php]: Isn't this a quite nice fail safe to have in place?

File: src/Hmac/Adapter/SymfonyRequestAdapter.php (hunk 1 above)
Comment (lordrhodos) [reply src/Hmac/Adapter/SymfonyRequestAdapter.php]: yes, just copied the method body from the symfony method, will change it.

File: tests/Mocks/Psr7Request.php [hunk 3]
```diff
@@ -0,0 +1,134 @@
+<?php declare(strict_types=1);
+
+namespace Starlit\Request\Authenticator\Tests\Mocks;
+
+use Psr\Http\Message\RequestInterface;
+use Psr\Http\Message\StreamInterface;
+use Psr\Http\Message\UriInterface;
+
+class Psr7Request implements RequestInterface
```
[the remark this answers is not in the document]
Comment (lordrhodos) [reply tests/Mocks/Psr7Request.php]: yes we can, but would need to define the return values of the required functions and couple the different mocks. Maybe still better than defining custom mocking classes?

[the remark this answers is not in the document]
Comment (felixsand) [reply tests/Mocks/Psr7Request.php]: Yes I think so. A bit less code and a bit more flexibility if we need to modify it in various ways for future unit tests 👍

File: tests/Hmac/HmacAuthenticatorTest.php (hunk 2 above)
Comment (lordrhodos) [reply tests/Hmac/HmacAuthenticatorTest.php]: it is not needed anymore, as the secret gets only injected into the `HmacGenerator` and not the `HmacAuthenticator`. You can find the [test for the HmacGenerator here](https://github.com/starweb/starlit-request-authenticator/pull/4/files#diff-9759148343d07e6135732c93304967d9R52)

Comment (felixsand) [reply tests/Hmac/HmacAuthenticatorTest.php]: Excellent! 👍

**Pull request merged**